lib/SILOptimizer/Utils/ConstantFolding.cpp

```diff
@@ -989,6 +1008,83 @@ case BuiltinValueKind::id:
     return B.createFloatLiteral(Loc, BI->getType(), TruncVal);
   }
 
+  case BuiltinValueKind::FPToSI: {
+    auto *V = dyn_cast<FloatLiteralInst>(Args[0]);
+    if (!V)
+      return nullptr;
+    APFloat fpVal = V->getValue();
+    auto *destTy = Builtin.Types[1]->castTo<BuiltinIntegerType>();
+
+    llvm::APSInt resInt(destTy->getFixedWidth(), /*isUnsigned=*/ false);
+    bool isExact = false;
+    APFloat::opStatus status =
+      fpVal.convertToInteger(resInt, APFloat::rmTowardZero, &isExact);
+
+    if (ResultsInError.hasValue()) {
+      if(status & APFloat::opStatus::opInvalidOp) {
+        SmallString<10> fpStr;
+        fpVal.toString(fpStr);
+        diagnose(M.getASTContext(), BI->getLoc().getSourceLoc(),
+                 diag::float_to_int_overflow, fpStr,
+                 destTy->getFixedWidth(), "signed");
+        ResultsInError = Optional<bool>(true);
+        return nullptr;
+      }
+    }
+    // Do not fold if any of the flags are set.
+    if (status != APFloat::opStatus::opOK &&
+        status != APFloat::opStatus::opInexact) {
+      return nullptr;
+    }
+    // The call to the builtin should be replaced with the constant value.
+    SILBuilderWithScope B(BI);
+    return B.createIntegerLiteral(BI->getLoc(), BI->getType(), resInt);
+  }
+
+  case BuiltinValueKind::FPToUI: {
+    auto *V = dyn_cast<FloatLiteralInst>(Args[0]);
+    if (!V)
+      return nullptr;
+    APFloat fpVal = V->getValue();
+
+    // fpVal should be non-negative. Otherwise, report an error.
+    if(ResultsInError.hasValue() && fpVal.isNegative() && !fpVal.isZero()) {
+      SmallString<10> fpStr;
+      fpVal.toString(fpStr);
+      diagnose(M.getASTContext(), BI->getLoc().getSourceLoc(),
+               diag::negative_fp_literal_overflow_unsigned, fpStr);
+      ResultsInError = Optional<bool>(true);
+      return nullptr;
+    }
+
+    auto *destTy = Builtin.Types[1]->castTo<BuiltinIntegerType>();
+
+    llvm::APSInt resInt(destTy->getFixedWidth(), /*isUnsigned=*/ true);
+    bool isExact = false;
+    APFloat::opStatus status =
+      fpVal.convertToInteger(resInt, APFloat::rmTowardZero, &isExact);
+
+    if (ResultsInError.hasValue()) {
+      if(status & APFloat::opStatus::opInvalidOp) {
+        SmallString<10> fpStr;
+        fpVal.toString(fpStr);
+        diagnose(M.getASTContext(), BI->getLoc().getSourceLoc(),
+                 diag::float_to_int_overflow, fpStr,
+                 destTy->getFixedWidth(), "unsigned");
+        ResultsInError = Optional<bool>(true);
+        return nullptr;
+      }
+    }
+    // Do not fold if any of the flags are set
```

Comment:
Does it make sense to factor this out with the previous occurrence of this logic?